CUDA Cooperative grid groups

gmarkall opened the pull request:
This picks up PR #4551(addition of cooperative grid groups and grid sync) and addresses the comments I made on that PR - the implementation is based on @njwhite's work with changes added. Changes on top of #4551 include:

* Added documentation.
* Cooperative launches are automatically inferred based on the use of grid groups in the kernel - there is no need for the user to manually specifying a cooperative launch.
* `cudadevrt` is linked as required, rather than needing to be manually located and linked by the user. `cudadevrt` is available in the Anaconda-packaged cudatoolkit now (https://github.com/AnacondaRecipes/cudatoolkit-feedstock/pull/5), so it is generally available where Numba is installed with conda.
* Additional tests added.
* `this_grid` is moved into the `cuda.cg` module, mirroring the `cg` namespace in CUDA C++. I expect more to be added to the `cg` module in future PRs (e.g. thread groups, etc.).

Some performance results when using CGs and grid sync to push a loop over matrix rows from the host into a kernel are discussed [in this comment](https://github.com/TDAmeritrade/stumpy/issues/245#issuecomment-692793221) - A summary:

```
N         Time (seconds)  Original time (s)
========  ==============  =================
  10000:            0.08               3.13
 100000:            2.21              31.97
1000000:          147.37             325.00
```

(`Time` is for the kernel with a loop and grid sync, and `Original time` is the loop in host code and multiple kernel launches)

seanlaw commented:
For completeness, [this comment may also be relevant](https://github.com/TDAmeritrade/stumpy/issues/245#issuecomment-693399894)

stuartarchibald commented on numba/cuda/tests/cudapy/test_cooperative_groups.py:
```diff
@@ -0,0 +1,152 @@
+from numba.cuda.testing import unittest, CUDATestCase, skip_on_cudasim
+
+
+def cc_X_or_above(major, minor):
+    if not config.ENABLE_CUDASIM:
+        cc = cuda.current_context().device.compute_capability
+        return cc >= (major, minor)
+    else:
+        return True
+
+
+def skip_unless_cc_60(fn):
+    return unittest.skipUnless(cc_X_or_above(6, 0), "requires cc >= 6.0")(fn)
```
Should these go into a `numba.cuda.tests.utils` or similar, there's been a few of these of late?

gmarkall replied:
There's similar things in `numba.cuda.testing`, so I've moved them there.

stuartarchibald commented on numba/cuda/tests/cudapy/test_cooperative_groups.py:
```diff
@@ -0,0 +1,152 @@
+    # The grid writes rows one at a time. Each thread reads an element from
+    # the previous row written by its "opposite" thread.
+    #
+    # A failure to sync the grid at each row would result in an incorrect
+    # result as some threads could run ahead of threads in other blocks, or
+    # fail to see the update to the previous row from their opposite thread.
```
I'm not sure that this is a would, more of a could? If you don't sync the grid across rows does it always fail, just wondering if there's something hardware-level that could cause occasional accidental success?

gmarkall replied:
I think of it as being so certain to happen, that I used the word "would" - the number of cycles that it would take for global memory writes by one block to be seen by another block is so large compared to the small number of cycles in the loop that it seems practically impossible for this kernel to ever complete its execution consistently - you'd have to get lucky for every element of every row in one run.

gmarkall replied:
> practically impossible for this kernel to ever complete its execution consistently

practically impossible for this kernel to ever complete its execution consistently **if the sync was omitted or didn't work**, that is :-)

stuartarchibald replied:
indeed :) ok, on the basis that it's spectacularly unlucky to hit this otherwise, and the context is about adding the thing to make it work, let's leave it :)

stuartarchibald commented on numba/cuda/tests/cudapy/test_cooperative_groups.py:
```diff
@@ -0,0 +1,152 @@
+        this_grid[1, 1](A)
+
+        # Ensure the kernel executed beyond the call to cuda.this_grid()
+        self.assertFalse(np.isnan(A[0]), 'set it to something!')
```
```suggestion
        self.assertFalse(np.isnan(A[0]), 'Value erroneously set!')
```

gmarkall replied:
I've added a slightly different phrase: "Value not set" - if `A[0]` is a NaN, then it was not set by the kernel.

stuartarchibald commented on numba/cuda/tests/cudapy/test_cooperative_groups.py:
```diff
@@ -0,0 +1,152 @@
+        A = np.full(1, fill_value=np.nan)
+        this_grid[1, 1](A)
+
+        # this_grid should have been determinted to be cooperative
```
```suggestion
        # this_grid should have been determined to be cooperative
```

gmarkall replied:
Fixed.

stuartarchibald commented on docs/source/cuda/cooperative_groups.rst:
```diff
@@ -0,0 +1,120 @@
+==================
+Cooperative Groups
```
Reminder from OOB chat to consider adding a note about Anaconda CTK version 10+ needed unless backport of static libs is done.

gmarkall replied:
Added.

stuartarchibald reviewed (commented):
Thanks for the patch, great to see this feature added. The patch looks good on the whole, just a few minor things to resolve.

gmarkall commented on numba/cuda/nvvmutils.py:
```diff
@@ -111,6 +111,20 @@ def declare_atomic_nanmin_float64(lmod):
     return lmod.get_or_insert_function(fnty, fname)
 
 
+def declare_cudaCGGetIntrinsicHandle(lmod):
+    fname = 'cudaCGGetIntrinsicHandle'
+    fnty = lc.Type.function(lc.Type.int(64),
+                            (lc.Type.int(32),))
+    return lmod.get_or_insert_function(fnty, fname)
+
+
+def declare_cudaCGSynchronize(lmod):
+    fname = 'cudaCGSynchronize'
+    fnty = lc.Type.function(lc.Type.int(32),
+                            (lc.Type.int(64), lc.Type.int(32)))
+    return lmod.get_or_insert_function(fnty, fname)
```
They're defined in `libcudadevrt.a`.

stuartarchibald replied:
Apologies, I was not sufficiently clear, I was looking for some documentation which specified the prototypes for these functions.

gmarkall replied:
There's no documentation, but the prototypes are in `include/cuda_device_runtime_api.h` in an installation:

```C++
extern __device__ __cudart_builtin__ unsigned long long CUDARTAPI cudaCGGetIntrinsicHandle(enum cudaCGScope scope);
extern __device__ __cudart_builtin__ cudaError_t CUDARTAPI cudaCGSynchronize(unsigned long long handle, unsigned int flags);
```

stuartarchibald replied:
thanks, sigs match up.

gmarkall commented on docs/source/cuda/cooperative_groups.rst:
```diff
@@ -0,0 +1,120 @@
+       for row in range(1, rows):
+           opposite = cols - col - 1
+           # Each row's elements are one greater than the previous row
+           M[row, col] = M[row - 1, opposite] + 1
+           # Wait until all threads have written their column element,
+           # and that the write is visible to all other threads
+           g.sync()
+
+   # Empty input data
+   A = np.zeros((1024, 1024), dtype=np.int32)
+   # A somewhat arbitrary choice (one warp), but generally smaller block sizes
+   # allow more blocks to be launched (noting that other limitations on
+   # occupancy apply such as shared memory size)
+   blockdim = 32
+   griddim = A.shape[1] // blockdim
+
+   # Kernel launch - this is implicitly a cooperative launch
+   sequential_rows[griddim, blockdim](A)
+
+   # Sanity check - are the results what we expect?
+   reference = np.tile(np.arange(1024), (1024, 1)).T
+   np.testing.assert_equal(A, reference)
+
+   # What do the results look like?
+   print(A)
+
+   # [[   0    0    0 ...    0    0    0]
+   #  [   1    1    1 ...    1    1    1]
+   #  [   2    2    2 ...    2    2    2]
+   #  ...
+   #  [1021 1021 1021 ... 1021 1021 1021]
+   #  [1022 1022 1022 ... 1022 1022 1022]
+   #  [1023 1023 1023 ... 1023 1023 1023]]
+
+The maximum grid size for ``sequential_rows`` can be enquired using:
+
+
+.. code-block:: python
+
+   defn = sequential_rows.definition
```
Thankyou! :-)

gmarkall commented on docs/source/cuda/cooperative_groups.rst:
```diff
@@ -0,0 +1,120 @@
+as a normal kernel launch - Numba automatically determines whether a cooperative
+launch is required based on whether a grid group is synchronized in the kernel.
+
+The grid size for a cooperative launch is more than for a normal launch - the
```
I intended to convery that it is more restrictive - somehow I must have messed up the grammar when editing.

gmarkall commented on docs/source/cuda/cooperative_groups.rst:
```diff
@@ -0,0 +1,120 @@
+Numba's Cooperative Groups support presently provides grid groups and grid
+synchronization, along with cooperative kernel launches.
+
+Cooperative groups are supported on Linux, and Windows for devices in TCC mode.
```
Added.

gmarkall commented on docs/source/cuda/cooperative_groups.rst:
```diff
@@ -0,0 +1,120 @@
+thread. A grid sync is needed to ensure that threads in the grid don't run ahead
+of threads in other blocks, or fail to see updates from their opposite thread.
+
+.. code-block:: python
```
This is now added as a doc test, using the same infrastructure for CUDA doc tests as added in #6430.

gmarkall commented:
@stuartarchibald Many thanks for the review - I believe I've now answered / addressed all comments.

stuartarchibald commented:
Buildfarm ID: `numba_smoketest_cuda_134`

stuartarchibald commented:
> @stuartarchibald Many thanks for the review - I believe I've now answered / addressed all comments.

No problem, thanks for the fixes, patch looks good!

stuartarchibald commented:
On the build farm this is producing numerous error messages like:
```
INFO - ERROR: test_sync_group_is_cooperative (numba.cuda.tests.cudapy.test_cooperative_groups.TestCudaCooperativeGroups)
INFO - ----------------------------------------------------------------------
INFO - Traceback (most recent call last):
INFO - File "<env>/lib/python3.8/site-packages/numba/cuda/tests/cudapy/test_cooperative_groups.py", line 81, in test_sync_group_is_cooperative
INFO - sync_group[1, 1](A)
INFO - File "<env>/lib/python3.8/site-packages/numba/cuda/compiler.py", line 799, in __call__
INFO - return self.dispatcher.call(args, self.griddim, self.blockdim,
INFO - File "<env>/lib/python3.8/site-packages/numba/cuda/compiler.py", line 906, in call
INFO - kernel.launch(args, griddim, blockdim, stream, sharedmem)
INFO - File "<env>/lib/python3.8/site-packages/numba/cuda/compiler.py", line 657, in launch
INFO - cufunc = self._func.get()
INFO - File "<env>/lib/python3.8/site-packages/numba/cuda/compiler.py", line 454, in get
INFO - linker.add_file_guess_ext(path)
INFO - File "<env>/lib/python3.8/site-packages/numba/cuda/cudadrv/driver.py", line 2185, in add_file_guess_ext
INFO - kind = FILE_EXTENSION_MAP[ext]
INFO - KeyError: 'so'
```
on all linux builds (CUDA 9 and CUDA 10). Some of those builds are CUDA 9 and might be expected to have issues, but the windows counterparts were ok.

gmarkall commented:
> On the build farm this is producing numerous error messages like:
> 
> on all linux builds (CUDA 9 and CUDA 10). Some of those builds are CUDA 9 and might be expected to have issues, but the windows counterparts were ok.

I'm guessing this is for CUDA 9.2, 10.0 and 10.1? The defaults channel conda cudatoolkit packages don't include `cudadevrt.a` prior to 10.2, which would cause a problem with CGs. I think the Windows builds might have been OK if the Windows machine has a GPU with CC < 6.0, as the problematic tests would have been skipped for that reason instead.

There were a few other logic errors that made the error message unhelpful, so I've added the following changes to handle a missing `cudadevrt.a` gracefully, and help the user understand what's going on:

- Fixed the extension determined for static libraries when the library isn't found in the Conda env (i.e. `.a` instead of `.so`, which is why the `KeyError` reported above was produced).
- Added a check for static libraries (just `cudadevrt` for now) so that its location (or the fact it's not found) is reported in the library test, e.g.:

```
$ python -c "from numba import cuda; cuda.cudadrv.libs.test()"
Finding cublas from System
	located at /usr/local/cuda/lib64/libcublas.so.11.2.0.252
	trying to open library...	ok
Finding cusparse from System
	located at /usr/local/cuda/lib64/libcusparse.so.11.1.1.245
	trying to open library...	ok
Finding cufft from System
	located at /usr/local/cuda/lib64/libcufft.so.10.2.1.245
	trying to open library...	ok
Finding curand from System
	located at /usr/local/cuda/lib64/libcurand.so.10.2.1.245
	trying to open library...	ok
Finding nvvm from System
	located at /usr/local/cuda/nvvm/lib64/libnvvm.so.3.3.0
	trying to open library...	ok
Finding cudart from System
	located at /usr/local/cuda/lib64/libcudart.so.11.0.221
	trying to open library...	ok
Finding cudadevrt from System
	located at /usr/local/cuda/lib64/libcudadevrt.a
Finding libdevice from System
	searching for compute_20...	ok
	searching for compute_30...	ok
	searching for compute_35...	ok
	searching for compute_50...	ok
```

- Added a message stating the path when a library that can't be linked.
- Added a decorator to skip tests when cudadevrt is missing, and use it on the CG tests.

Hopefully this should resolve things - could it have another buildfarm run please (assuming the Azure CI is OK)?

stuartarchibald commented:
Thanks for fixing this up @gmarkall. Looks like Azure is failing and in a legitimate way, seems like it could be something minor causing it. Will push this through the farm again once Azure is passing.

gmarkall commented:
@stuartarchibald I think the latest fail was just a network error - could you re-run the CI please?

esc commented:
/azp run

gmarkall commented:
Thanks for the run :-)

Unfortunately it now looks like the llvmlite version bumping is causing some issues :-)

stuartarchibald commented:
/AzurePipelines run

stuartarchibald reviewed (approved):
Thanks for the fixes, additional code changes look good, will give it a run through the farm.

stuartarchibald commented:
Buildfarm ID: `numba_smoketest_cuda_136`

stuartarchibald commented:
> Buildfarm ID: `numba_smoketest_cuda_136`

Passed.
